### ISSUE TITLE
Inconsistency when signing data

### ISSUE THREAD

**Issue by michaeljgea** [nagyesta/lowkey-vault#651]
First of all, thanks @nagyesta for this awesome project. 

#### Describe the bug

When using the official client and connecting to a real key vault, I cannot use the `CryptographyClient#sign` method to sign an arbitrary amount of data. It wants you to already send a digest compatible with the specified algorithm. For example, if I choose SHA256, it wants a byte[] with exactly 32 bytes. The `signData` method on the other hand accepts any kind of input for the `data` parameter. In this case, the CryptographyClient builds the digest for you.

However, in the lowkeyvault client, the `CryptographyClient#sign` allows arbitrary data as the method argument and it will digest it. In short, the real keyvault's `signData` method is functionally equivalent with lowkeyvault's `sign` method. 

The `signData` method of lowkeyvault on the other hand seems to do something entirely different. At least in my tests, it did not create the signatures I expected.

#### To reproduce

For the `sign` method:
1. Call the `sign` method of the real key vault with a byte[] of size 33 -> Exception
2. Call the `sign` method of lowkey vault vault with a byte[] of size 33 -> Creates a signature for the digest of the given byte array.

For the `signData` method:
- Compare the result of calling the `signData` method of the real key vault and the lowkey vault. They differ.

#### Environment

Version: 1.28.94

**Comment by nagyesta** [comment]
Hi @michaeljgea ,

Thank you for giving Lowkey Vault a chance and thank you for the great quality bug report! I wasn't aware of how these endpoints worked exactly, so I think I didn't add any restrictions around them. 

I need to take a better look at how the client generates the requests and what is the source of the difference you have seen between the official and the Lowkey Vault clients, considering that the Lowkey Vault client is only supposed to configure an instance of the official client for you. Anyway, I will look into it for sure and figure out what the cause might be.

How urgently do you need the fix?

**Comment by michaeljgea** [comment]
It's not urgent. We have a workaround in place where we switch the signing method used with a lambda, based on whether or not we are using the real keyvault or lowkey vault as a backend. Obviously, it would be optimal to have lowkey vault as a drop-in replacement, but it is more of a nice-to-have ;-)

**Comment by michaeljgea** [comment]
If it helps, I can also try to provide a minimum viable example code to reproduce the problem.

**Comment by nagyesta** [comment]
Hi @michaeljgea , 
sorry for the slow response! The example you have suggested would be a huge help, thank you in advance!
If you could share the example code as well as what kind of key I should create in both the real key vault and in Lowkey Vault, then I would see the issue easily! 

I can most probably take a look over the weekend and figure out how hard it will be to fix it.

**Comment by michaeljgea** [comment]
Here is an example:
```
package com.gea.ft.service.enrollment.controller;

import com.azure.core.util.polling.SyncPoller;
import com.azure.identity.DefaultAzureCredential;
import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.security.keyvault.certificates.CertificateClient;
import com.azure.security.keyvault.certificates.CertificateClientBuilder;
import com.azure.security.keyvault.certificates.models.CertificateContentType;
import com.azure.security.keyvault.certificates.models.CertificateKeyType;
import com.azure.security.keyvault.certificates.models.CertificateKeyUsage;
import com.azure.security.keyvault.certificates.models.CertificateOperation;
import com.azure.security.keyvault.certificates.models.CertificatePolicy;
import com.azure.security.keyvault.certificates.models.KeyVaultCertificateWithPolicy;
import com.azure.security.keyvault.keys.KeyClient;
import com.azure.security.keyvault.keys.KeyClientBuilder;
import com.azure.security.keyvault.keys.cryptography.CryptographyClient;
import com.azure.security.keyvault.keys.cryptography.models.SignResult;
import com.azure.security.keyvault.keys.cryptography.models.SignatureAlgorithm;
import com.github.nagyesta.lowkeyvault.http.ApacheHttpClientProvider;
import com.github.nagyesta.lowkeyvault.http.AuthorityOverrideFunction;
import com.github.nagyesta.lowkeyvault.testcontainers.LowkeyVaultContainer;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;
import org.testcontainers.images.PullPolicy;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;
import org.testcontainers.utility.DockerImageName;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.security.PublicKey;
import java.security.Signature;
import java.security.cert.CertificateFactory;
import java.util.Collections;
import java.util.HashSet;
import java.util.UUID;

import static com.github.nagyesta.lowkeyvault.testcontainers.LowkeyVaultContainerBuilder.lowkeyVault;
import static org.assertj.core.api.Assertions.assertThat;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
@Testcontainers
@Slf4j
public class MveTest {
    private static final String LOWKEY_VAULT_NAME = "default";
    private static final String ROOT_CA_NAME = "CN=my-root-ca";
    private static final String LOREM_IPSUM =
            "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";
    private static final String REAL_KEY_VAULT_URL = "https://FILL-ME.vault.azure.net/";

    @Container
    static LowkeyVaultContainer vault = lowkeyVault(DockerImageName.parse("nagyesta/lowkey-vault:1.28.94")).vaultNames(
            new HashSet<>() {
                {
                    this.add(LOWKEY_VAULT_NAME); // Set.of(..) did not work somehow
                }
            }).build().withImagePullPolicy(PullPolicy.defaultPolicy());

    @Test
    void given_lowkey_vault_when_signing_long_text_with_sign_method() {
        // generate root certificate
        ApacheHttpClientProvider clientProvider = new ApacheHttpClientProvider(vault.getVaultBaseUrl(LOWKEY_VAULT_NAME),
                new AuthorityOverrideFunction(vault.getVaultAuthority(LOWKEY_VAULT_NAME), vault
                        .getEndpointAuthority()));
        CertificateClient certificateClient = clientProvider.getCertificateClient();
        KeyClient keyClient = clientProvider.getKeyClient();
        KeyVaultCertificateWithPolicy rootCertificate = generateRootCertificate(certificateClient);
        PublicKey rootCertificatePK = getPublicKey(rootCertificate.getCer());
        CryptographyClient cryptographyClient = getCryptographyClient(keyClient, rootCertificate);

        SignResult result = cryptographyClient.sign(SignatureAlgorithm.RS256, LOREM_IPSUM.getBytes());

        assertThat(checkSignature(rootCertificatePK, result.getSignature())).isTrue();
    }

    @Test
    void given_lowkey_vault_when_signing_long_text_with_signData_method() {
        // generate root certificate
        ApacheHttpClientProvider clientProvider = new ApacheHttpClientProvider(vault.getVaultBaseUrl(LOWKEY_VAULT_NAME),
                new AuthorityOverrideFunction(vault.getVaultAuthority(LOWKEY_VAULT_NAME), vault
                        .getEndpointAuthority()));
        CertificateClient certificateClient = clientProvider.getCertificateClient();
        KeyClient keyClient = clientProvider.getKeyClient();
        KeyVaultCertificateWithPolicy rootCertificate = generateRootCertificate(certificateClient);
        PublicKey rootCertificatePK = getPublicKey(rootCertificate.getCer());
        CryptographyClient cryptographyClient = getCryptographyClient(keyClient, rootCertificate);

        SignResult result = cryptographyClient.signData(SignatureAlgorithm.RS256, LOREM_IPSUM.getBytes());

        assertThat(checkSignature(rootCertificatePK, result.getSignature())).isTrue();
    }

    @Test
    void given_real_vault_when_signing_long_text_with_sign_method() {
        // generate root certificate
        DefaultAzureCredential credentials = new DefaultAzureCredentialBuilder().build();
        CertificateClient certificateClient = new CertificateClientBuilder().vaultUrl(REAL_KEY_VAULT_URL)
                .credential(credentials)
                .buildClient();
        KeyClient keyClient = new KeyClientBuilder().vaultUrl(REAL_KEY_VAULT_URL).credential(credentials).buildClient();
        KeyVaultCertificateWithPolicy rootCertificate = generateRootCertificate(certificateClient);
        PublicKey rootCertificatePK = getPublicKey(rootCertificate.getCer());
        CryptographyClient cryptographyClient = getCryptographyClient(keyClient, rootCertificate);

        SignResult result = cryptographyClient.sign(SignatureAlgorithm.RS256, LOREM_IPSUM.getBytes());

        assertThat(checkSignature(rootCertificatePK, result.getSignature())).isTrue();
    }

    @SneakyThrows
    @Test
    void given_real_vault_when_signing_long_text_with_signData_method() {
        // generate root certificate
        DefaultAzureCredential credentials = new DefaultAzureCredentialBuilder().build();
        CertificateClient certificateClient = new CertificateClientBuilder().vaultUrl(REAL_KEY_VAULT_URL)
                .credential(credentials)
                .buildClient();
        KeyClient keyClient = new KeyClientBuilder().vaultUrl(REAL_KEY_VAULT_URL).credential(credentials).buildClient();
        KeyVaultCertificateWithPolicy rootCertificate = generateRootCertificate(certificateClient);
        PublicKey rootCertificatePK = getPublicKey(rootCertificate.getCer());
        CryptographyClient cryptographyClient = getCryptographyClient(keyClient, rootCertificate);

        SignResult result = cryptographyClient.signData(SignatureAlgorithm.RS256, LOREM_IPSUM.getBytes());

        assertThat(checkSignature(rootCertificatePK, result.getSignature())).isTrue();
    }

    @SneakyThrows
    private static boolean checkSignature(PublicKey publicKey, byte[] signatureToCheck) {
        Signature sig = Signature.getInstance("SHA256WithRSA");
        sig.initVerify(publicKey);
        sig.update(LOREM_IPSUM.getBytes());
        return sig.verify(signatureToCheck);
    }

    private CryptographyClient getCryptographyClient(KeyClient keyClient,
                                                     KeyVaultCertificateWithPolicy rootCertificate) {
        String[] keyIdSegments = rootCertificate.getKeyId().split("/");
        String keyName = keyIdSegments[keyIdSegments.length - 2];
        String keyVersion = keyIdSegments[keyIdSegments.length - 1];
        return keyClient.getCryptographyClient(keyName, keyVersion);
    }

    private KeyVaultCertificateWithPolicy generateRootCertificate(CertificateClient certificateClient) {
        CertificatePolicy policy = new CertificatePolicy("Self", ROOT_CA_NAME);
        policy.setExportable(false);
        policy.setKeySize(2048);
        policy.setKeyType(CertificateKeyType.RSA);
        policy.setContentType(CertificateContentType.PKCS12);
        policy.setValidityInMonths(300);
        policy.setKeyReusable(false);
        policy.setKeyUsage(CertificateKeyUsage.KEY_CERT_SIGN,
                CertificateKeyUsage.CRL_SIGN,
                CertificateKeyUsage.DIGITAL_SIGNATURE);
        String certName = UUID.randomUUID().toString();
        SyncPoller<CertificateOperation, KeyVaultCertificateWithPolicy> createOperation = certificateClient
                .beginCreateCertificate(certName, policy, true, Collections.emptyMap());
        createOperation.waitForCompletion();
        return certificateClient.getCertificate(certName);
    }

    @SneakyThrows
    private PublicKey getPublicKey(byte[] certificate) {
        CertificateFactory cf = CertificateFactory.getInstance("X.509");
        InputStream in = new ByteArrayInputStream(certificate);
        return cf.generateCertificate(in).getPublicKey();
    }
}
```

Dependencies in the pom:
- lombok
- Testcontainers
- Junit 5
- assertj
- azure-security-keyvault-certificates
- azure-security-keyvault-keys
- azure-core
- azure-identity
- Lowkeyvault obviously :-D

To test it on your side, you have to set the `REAL_KEY_VAULT_URL` constant to a real key vault, where your Azure CLI user has the permissions GET and LIST for certificates and GET, LIST, and SIGN for keys.

The results on my end are as expected:
- `given_real_vault_when_signing_long_text_with_sign_method` -> Exception, byte[] parameter should be 32 bytes in base64url encoding
- `given_lowkey_vault_when_signing_long_text_with_sign_method` -> Succeeds (which is not consistent with the test above)
- `given_real_vault_when_signing_long_text_with_signData_method` -> Succeeds 
- `given_lowkey_vault_when_signing_long_text_with_signData_method` -> Produces a signature, which cannot be verified against the public key of the signing certificate.
  - One side note here: When I use the lowkey vault's built-in verify method instead of `java.security.Signature`, then this test succeeds, which suggests that the signData and verify methods in lowkey vault are both implemented wrongly, but in a way that when applied together, they still work. Of course, this would not be helpful in actual use cases, where you either sign or verify something but not both.
  
Hope this helps!

**Comment by nagyesta** [comment]
Thank you @michaeljgea this is great! I will look into it over the weekend and see what I can do about it!

**Comment by nagyesta** [comment]
The inconsistency around ```signData``` must be because double hashing. The ```signData``` call is performing a ```SHA256``` hashing even before the service is called. Then inside Lowkey Vault, the ```SHA256withRSA``` algorithm is used, which will hash the digest then sign as far as I understand. I will continue the investigation tomorrow, but this is promising.

**Comment by nagyesta** [comment]
I have managed to verify this with all available key types and the supported algorithms. Now, I can see, that `EC` cases are working as expected, but all `RSA` algorithms are buggy (behaving different to the real key vault). 

I know how to fix the `RS256`, `RS384`, `RS512` cases, but I have some trouble around the `PS256`, `PS384`, `PS512`, because I would need to let the client do the `SHA` hashing and then perform the `MGF1` part with the signing (which I don't know how to do yet as I am not that strong in cryptography). Anyway, I will continue working on this and hopefully figure out how it can be done. 

Worst case scenario, I might end up splitting out the `MGF1` part into another bug and fix only the rest here.

**Comment by michaeljgea** [comment]
Thanks for the quick investigation 👍

**Comment by nagyesta** [comment]
Hi @michaeljgea ,

thank you for all the help you have provided during this fix! The updated code was just released in `v1.29.0`. It should show up in Maven Central in a couple of minutes (and in searches in 1-2 hours max). The main Docker artifact is already available and the ARM build will be released as soon as Renovate notices the new Jar version.

Best of luck and please let me know in case you see any issues!

Thank you!

**Comment by michaeljgea** [comment]
I can confirm that (at least for SHA256) the sign and signData methods are now behaving as expected. 

Amazing work! Thanks a bunch.
